pkg/declarative/v2/reconciler.go

```diff
@@ -0,0 +1,382 @@
+	if err := r.Get(ctx, req.NamespacedName, obj); err != nil {
+		logger.Info(req.NamespacedName.String() + " got deleted!")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if r.ShouldSkip(ctx, obj) {
+		return ctrl.Result{}, nil
+	}
+
+	// ServerSideApply requires nil ManagedFields
+	obj.SetManagedFields(nil)
+
+	status := obj.GetStatus()
+
+	if !obj.GetDeletionTimestamp().IsZero() && obj.GetStatus().State != StateDeleting {
+		obj.SetStatus(status.WithState(StateDeleting).WithOperation("switched to deleting"))
+		return r.ssaStatus(ctx, obj)
+	}
+
+	// add finalizer
+	if controllerutil.AddFinalizer(obj, r.Finalizer) {
+		return r.ssa(ctx, obj)
+	}
+
+	var (
+		resourceCondition = metav1.Condition{
+			Type:               string(ConditionTypeResources),
+			Reason:             string(ConditionReasonResourcesAreAvailable),
+			Status:             metav1.ConditionFalse,
+			Message:            "resources are parsed and ready for use",
+			ObservedGeneration: obj.GetGeneration(),
+		}
+		installationCondition = metav1.Condition{
+			Type:               string(ConditionTypeInstallation),
+			Reason:             string(ConditionReasonReady),
+			Status:             metav1.ConditionFalse,
+			Message:            "installation is ready and resources can be used",
+			ObservedGeneration: obj.GetGeneration(),
+		}
+	)
+
+	// Processing
+	if status.State == "" {
+		presetConditions := 2
+		status.Conditions = make([]metav1.Condition, 0, presetConditions)
+		meta.SetStatusCondition(&status.Conditions, resourceCondition)
+		meta.SetStatusCondition(&status.Conditions, installationCondition)
+		status.Synced = make([]Resource, 0)
+		obj.SetStatus(status.WithState(StateProcessing).WithOperation("switched to processing"))
+		return r.ssaStatus(ctx, obj)
+	}
+
+	spec, err := r.Spec(ctx, obj)
+	if err != nil {
+		r.Event(obj, "Warning", "Spec", err.Error())
+		obj.SetStatus(status.WithState(StateError).WithErr(err))
+		return r.ssaStatus(ctx, obj)
+	}
+
+	skr, err := r.initializeClient(ctx, obj, spec)
+	if err != nil {
+		r.Event(obj, "Warning", "ClientInitialization", err.Error())
+		obj.SetStatus(status.WithState(StateError).WithErr(err))
+		return r.ssaStatus(ctx, obj)
+	}
+	resourceConverter := NewResourceConverter(skr, r.Namespace)
+
+	var renderer Renderer
+
+	switch spec.Mode {
+	case RenderModeHelm:
+		renderer = NewHelmRenderer(spec, skr, r.Options)
+		renderer = WrapWithRendererCache(renderer, spec, r.Options)
+	case RenderModeKustomize:
+		renderer = NewKustomizeRenderer(spec, r.Options)
+		renderer = WrapWithRendererCache(renderer, spec, r.Options)
+	case RenderModeRaw:
+		renderer = NewRawRenderer(spec, r.Options)
+	}
+
+	if err := renderer.Initialize(obj); err != nil {
+		return r.ssaStatus(ctx, obj)
+	}
+
+	if err := renderer.EnsurePrerequisites(ctx, obj); err != nil {
+		return r.ssaStatus(ctx, obj)
+	}
+
+	var target, current kube.ResourceList
+
+	if obj.GetDeletionTimestamp().IsZero() {
+		var targetResources *types.ManifestResources
```

Comment:
This If block can extract to some function like `getTargetListFromManifest`

Reply:
I was thinking about extracting this but ultimately decided against it in favour of easier readability of the reconcile loop from top-to-bottom. When I extracted it, it was a bit harder to understand how the actual render happens. If you disagree I will try to extract it nevertheless.

Reply:
OK, understood, I think then it's just a habit issue. for me, reading code is prefer to know based on level of abstraction, for example here in `Reconcile` function loop, I actually don't expected to know much detail about how the target get converted but simply in this condition, a list of target get generated, then I can skip and continue reading later logic in `Reconcile` how to deal with this `target`, but not distracted for this big block. After finish reading  `Reconcile`, then I jump into each function to know detail. If this not fit your pattern, it's fine leave like this to me.

Reply:
Understood, I think the point is valid. I will try to extract it and push a commit when I found a suitable fit or update the comment here then.

Reply:
extracted and split, I think we should continue writing unit tests on this